### ISSUE TITLE
Automatically add containers to no_proxy

### ISSUE THREAD

**Issue by krakhead22** [docker/compose#11433]
### Description

When working with a proxy configured in docker config.json it automatically passes the proxy information to the created containers. The problem is that some projects using docker compose refer to containers via their hostname.  With a proxy configured these connections can be sent to the proxy server rather than the expected container.

Since coordinating between multiple containers is most likely to occur within a compose project, and compose should be able to determine which containers are networked it would be nice if entries were automatically added to the NO_PROXY environment variables within the containers.  

An example project where I encountered this behavior is https://github.com/ckan/ckan-docker which builds & runs fine out of the box on a non-proxied connection, but when a proxy is enabled ckan can't reach solr. I was able to add a few lines to the startup scripts to get past my issue, but it seems like a universal solution is feasible and would be better.

**Comment by ndeloof** [comment]
Within a compose application you access other services using service alias, don't you?
So you can just configure your compose file with `environment` set to declare `NO_PROXY` accordingly:
```yaml
services:
  db: 
    (...)
  app:
    image: backend
    environment:
      - NO_PROXY=db
```

**Comment by krakhead22** [comment]
You are correct, adding the environment variable to the compose file solves the problem for this instance, but my point is that it would be nice if projects could work on proxy connection without having to modify the compose files (assuming the proxy was configured with docker). Docker provides a mechanism to populate proxy information to containers automatically, it would be nice if compose could complete the job so the user doesn't have to override/augment the `NO_PROXY` variable.

**Comment by ja-pavi** [comment]
Hi I'm a UT Austin, and would like to work on this issue if possible!

**Comment by dwaynebradley** [comment]
I just ran into this recently myself. It was a pain to determine what was going on. DNS and proxies...it is always one or the other... 😂

@mikesir87, I like your work around for Docker but just as an FYI, Podman already does something very similar to what you mentioned. Each service automatically gets an alias of `<service-name>.dns.podman` so it is very easy to add `dns.podman` to `no_proxy` and `NO_PROXY` to avoid this situation. Would be nice if Docker did something automatically like this as well. That way we do not have to modify compose files to accommodate this.

Little example:
- Create a new network with Podman: `podman network create test-net`
- Run a container (Postgres in this example): `podman run --rm --name=db --network=test-net -p 5555:5432 -e POSTGRES_PASSWORD=idontcare postgres:latest`
- Start another container on the same network: `podman run --rm -ti --network=test-net alpine:latest /bin/sh`
- From inside that container you can ping both `db` and `db.dns.podman`

```
/ # ping db
PING db (10.89.0.2): 56 data bytes
64 bytes from 10.89.0.2: seq=0 ttl=42 time=0.463 ms
64 bytes from 10.89.0.2: seq=1 ttl=42 time=0.056 ms
64 bytes from 10.89.0.2: seq=2 ttl=42 time=0.056 ms
^C
--- db ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.056/0.191/0.463 ms
/ # ping db.dns.podman
PING db.dns.podman (10.89.0.2): 56 data bytes
64 bytes from 10.89.0.2: seq=0 ttl=42 time=0.040 ms
64 bytes from 10.89.0.2: seq=1 ttl=42 time=0.054 ms
64 bytes from 10.89.0.2: seq=2 ttl=42 time=0.053 ms
^C
--- db.dns.podman ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.040/0.049/0.054 ms
```
